app/src/main/java/com/github/se/eduverse/viewmodel/DashboardViewModel.kt

```diff
@@ -0,0 +1,61 @@
+package com.github.se.eduverse.viewmodel
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.github.se.eduverse.model.CommonWidgetType
+import com.github.se.eduverse.model.Widget
+import com.github.se.eduverse.repository.DashboardRepository
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.launch
+
+open class DashboardViewModel(private val dashboardRepository: DashboardRepository) : ViewModel() {
+
+  private val _widgetList = MutableStateFlow<List<Widget>>(emptyList())
+  open val widgetList: StateFlow<List<Widget>>
+    get() = _widgetList
+
+  open fun fetchWidgets(userId: String) {
+    viewModelScope.launch {
+      dashboardRepository
+          .getWidgets(userId)
+          ?.catch { e ->
+            Log.e("DashboardViewModel", "Error fetching widgets", e)
+            _widgetList.value = emptyList()
+          }
+          ?.collect { widgets -> _widgetList.value = widgets }
+          ?: run {
+            Log.e("DashboardViewModel", "getWidgets returned null")
+            _widgetList.value = emptyList()
+          }
+    }
+  }
+
+  open fun addWidget(widget: Widget) {
+    viewModelScope.launch {
+      dashboardRepository.addWidget(widget)
+      Log.d("DashboardViewModel", "Widget added: ${widget.widgetId}")
+    }
+  }
+
+  open fun removeWidget(widgetId: String) {
+    viewModelScope.launch {
+      dashboardRepository.removeWidget(widgetId)
+      Log.d("DashboardViewModel", "Widget removed: $widgetId")
+    }
+  }
+
+  open fun getCommonWidgets(): List<Widget> {
```

Comment:
I don't see the point in having this function as you end up updating the fields when you create the widgets, it's better to directly create a new Widget instance directly in DashboardScreen